Add Java codemod development tutorial (#9)
pixee/codemodder-docs: docs/languages/java/write_codemod.md

---
sidebar_position: 3
title: Writing our First Codemod
---

Now, let's write our first codemod! 

> The newer methods in `java.nio.file.Files` replaced the need for some of community-loved APIs in `org.apache.commons.io.FileUtils`. So, let's write a codemod to move `org.apache.commons.io.FileUtils#readLines()` to `java.nio.file.Files.readAllLines()`. The diffs should look something like this:

Reviewer: > some of community-loved APIs

Feels like this is missing a word or needs clarification; I'm not quite sure what this means.

```diff
- import org.apache.commons.io.FileUtils; // remove the import if possible
+ import java.nio.file.Files;

...

- List<String> lines = FilesUtils.readLines(file);
+ List<String> lines = Files.readAllLines(file.toPath());
```

## Step #1: Write the Semgrep to find the calls we want to change

> The first thing we need is some Semgrep to find all the invocations of `FileUtils#readLines()`. The [Semgrep Playground](https://semgrep.dev/playground/new) is a great place to iterate on new Semgrep queries. They have [plenty of docs](https://semgrep.dev/docs/) that we won't try to re-create here, so let's hand-wave a bit here and show you the finished product:

Reviewer: ```suggestion
The first thing we need is a Semgrep query to find all the invocations of `FileUtils#readLines()`. The [Semgrep Playground](https://semgrep.dev/playground/new) is a great place to iterate on new Semgrep queries. They have [plenty of docs](https://semgrep.dev/docs/) that we won't try to re-create here, so let's just show you the finished product:
```

```yaml
rules:
  - id: migrate-files-commons-io-to-nio-read-file-to-lines
    pattern-either:
      - pattern: (org.apache.commons.io.FileUtils).readLines($X)
      - pattern: (org.apache.commons.io.FileUtils).readLines($X, (Charset $Y))
```

## Step #2: Choose a type to implement

Here are some of the key types for creating a new codemod.

| Type                                            | Description                                                                         |
|-------------------------------------------------|-------------------------------------------------------------------------------------|
| io.codemodder.CodeChanger                       | The base `interface` for a codemod. Should rarely be implemented directly.          |
| io.codemodder.javaparser.JavaParserChanger      | A codemod that gets handed a fully-built JavaParser model of a source file.         |
| io.codemodder.javaparser.SarifJavaParserChanger | A codemod that reads SARIF results and acts on their matching JavaParser AST nodes. |

> So -- what is SARIF? The [Static Analysis Results Interchange Format (SARIF)](https://sarifweb.azurewebsites.net/) is a JSON format that static analysis tools use to capture their results. You don't need to know too much (anything?) about SARIF -- but we're going to use `SarifJavaParserChanger` because that's the type that takes SARIF results from Semgrep and feeds them to JavaParser.

Reviewer: > You don't need to know too much (anything?) about SARIF

Just a comment but that suggests an opportunity for abstraction.

Author: Definitely

## Step #3: Code it up

We're going to step through the codemod from the top to the bottom, piece by piece, then show you the whole thing.

### The Class Definition

Let's look at the class definition and meet a few new concepts

```java
@Codemod(
    id = "pixee:java/migrate-files-commons-io-to-nio",
    reviewGuidance = ReviewGuidance.MERGE_WITHOUT_REVIEW)
public final class ReadLinesCodemod extends SarifPluginJavaParserChanger<MethodCallExpr> {
  ...
}
```

> Each codemod needs a `@Codemod` annotation. This gives some necessary metadata for end users and will eventually feed a more robust CLI experience.

Reviewer: ```suggestion
First, each codemod needs a `@Codemod` annotation. This gives some necessary metadata for end users and will eventually feed a more robust CLI experience.
```

> Second, notice the generic component of `SarifPluginJavaParserChanger`. This generic tells our framework that the Semgrep query returns the locations of Java method call expressions. If the SARIF points to a location that doesn't point to a method call, the callback won't occur. Let's see how to inject the Semgrep SARIF into the type.

Reviewer: ```suggestion
Second, notice the generic component of `SarifPluginJavaParserChanger`. This generic tells our framework that the Semgrep query returns the locations of Java method call expressions. If the SARIF points to a location that doesn't point to a method call, the callback won't execute. Let's see how to inject the Semgrep SARIF into the type.
```

              - pattern: (org.apache.commons.io.FileUtils).readLines($X, (Charset $Y))
        """;

  public ReadLinesCodemod(@SemgrepScan(yaml = RULE) final RuleSarif sarif) {
     super(
        sarif, // give it the SARIF that will be used to find hits
        MethodCallExpr.class, // only look for method calls that match the SARIF
        CodemodReporterStrategy.fromClasspath(ReadLinesCodemod.class) // pull storytelling text from classpath resource
     );
  }

  @Override
  public boolean onResultFound(
        final CodemodInvocationContext context,
        final CompilationUnit cu,
        final MethodCallExpr apacheReadLinesCall,
        final Result result) {
    NodeList<Expression> arguments = apacheReadLinesCall.getArguments();
    MethodCallExpr toPath = new MethodCallExpr(arguments.get(0), "toPath");
    arguments.set(0, toPath);
    boolean success =
        replace(apacheReadLinesCall)
            .withStaticMethod("java.nio.file.Files", "readAllLines")
            .withSameArguments();
    
    if (success) {
      removeImportIfUnused(cu, "org.apache.commons.io.FileUtils");
    }
    
    return success;
  }
}
```

## Step #3: Add the storytelling

> Well, we lied a little. We wrote all the _code_, but one of the philosophies of codemodder is that the storytelling around the changes we make has to be good. Without good storytelling, people won't know the purpose of value of the changes being made.

Reviewer: ```suggestion
Well, we lied a little. We wrote all the _code_, but one of the philosophies of codemodder is that we need to tell a good story about the changes we make. Without good storytelling, people won't know the purpose or value of the changes being made.
```

### Provide Storytelling From the Classpath (Recommended)

> We offered a shortcut for folks who don't want to main big blocks of text in a Java file (like us.) You can provide a [CodemodReporterStrategy](https://www.javadoc.io/doc/io.codemodder/codemodder-base/latest/io/codemodder/CodemodReporterStrategy.html) that reports based on text from a predictable location on the classpath. This is an alternative to storing data inline to the Java source code of your codemod. It's easier to maintain this "data" outside of code, so we prefer a simple mechanism for doing that. Both the files read are expected to be in `/com/acme/MyCodemod/` (assuming that's the name of your codemod type.)

Reviewer: ```suggestion
We provide a for folks who don't want to write and maintain big blocks of text in a Java file (like us). You can provide a [CodemodReporterStrategy](https://www.javadoc.io/doc/io.codemodder/codemodder-base/latest/io/codemodder/CodemodReporterStrategy.html) that builds reports based on text from a well-defined location on the classpath. This is an alternative to storing data inline to the Java source code of your codemod. It's easier to maintain this "data" outside of code, so we prefer a simple mechanism for doing that. Both the files read are expected to be in `/com/acme/MyCodemod/` (assuming that's the name of your codemod type).
```

 
The first expected file in that directory is {@code report.json}. It contains most of the fields we want to report:

```json
{
  "summary": "A headline describing the changes provided by this codemod",
  "change": "A description of the change suitable for a particular instance of a change",
  "references": [
    "A URL for further reading",
    "Another URL for further reading"
  ]
}
```

> The second file is `description.md`, and it provides last field needed by a `CodemodReporterStrategy` is the description of the codemod itself. This is expected to be a bigger piece of text, and thus it is stored in a separate file where it can be easily edited with an IDE supporting markdown.

Reviewer: > it provides last field

Not sure what this means, seems like a typo.

> The second file is `description.md`, and it provides last field needed by a `CodemodReporterStrategy` is the description of the codemod itself. This is expected to be a bigger piece of text, and thus it is stored in a separate file where it can be easily edited with an IDE supporting markdown.

Reviewer: ```suggestion
The second file is `description.md`, and it provides last field needed by a `CodemodReporterStrategy` is the description of the codemod itself. This is expected to contain a more detailed description so it is stored in a separate file where it can be easily edited with an IDE supporting markdown.
```

> Thus, in a typical Java project, using this `CodemodReporterStrategy` would mean your artifact would retain at least these 3 files:

Reviewer: ```suggestion
Thus, in a typical Java project, using this `CodemodReporterStrategy` would mean your artifact would include at minimum the following three files:
```

 - src/main/java/com/acme/MyCodemod.java
 - src/main/resources/com/acme/MyCodemod/report.json
 - src/main/resources/com/acme/MyCodemod/description.md

### Provide Storytelling In The Codemod
Alternatively, if you have very simple storytelling, you can directly implement the following methods in your codemod:

```java
@Override
public String getSummary() { return "Fixes Acme thing..."; }

@Override
public String getDescription() { return "This change ..."; }

@Override
public List<String> getReferences() { return List.of("https://internal.acme.com/use-x-instead-of-y"); }
```

### Skip Storytelling
If you don't care about storytelling, you can just use `CodemodReporterStrategy.empty()`.

## Back to our example

> So, for our codemod we're building, we'll choose the classpath reporter strategy, so we're going to add those two files:

Reviewer: ```suggestion
For our codemod, we'll choose the classpath reporter strategy, so we're going to add those two files:
```
